### PR TITLE
Add functional coverage test file 

### PR DESCRIPTION
**Pull request by hodjat91**
Add functional coverage test file.

### CODE REVIEW

File: pygen/pygen_src/test/riscv_instr_cov_test.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                                if not expect_illegal_instr:
+                                    logging.error("Found unexpected illegal instr: {} [{}]".format(self.trace["instr"], entry))
+                                    self.unexpected_illegal_instr_cnt += 1
+                            # TODO: call sample here to do the functional coverage, and move assign_trace_info into it
+                            #self.sample(instruction)
+                        self.entry_cnt += 1
+                    line_count += 1
+                logging.info("[{}]: {} instruction processed".format(csv_file, self.entry_cnt))
+                self.total_entry_cnt += self.entry_cnt
+        logging.info("Finished processing {} trace CSV, {} instructions".format(len(self.trace_csv), self.total_entry_cnt))
```
Comment (udinator) [inline pygen/pygen_src/test/riscv_instr_cov_test.py:84]: nit: here and in a few other places, keep line length to 80

Comment (hodjat91) [reply pygen/pygen_src/test/riscv_instr_cov_test.py:84]: Thank Udi, I will fix it.

File: pygen/pygen_src/test/riscv_instr_cov_test.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                               "lr" in self.trace["instr"] or \
+                               "sc" in self.trace["instr"]:
+                                # TODO: Enable functional coverage for AMO test
+                                pass
+                            if not self.sample():
+                                if not expect_illegal_instr:
+                                    logging.error("Found unexpected illegal instr: {} [{}]".format(self.trace["instr"], entry))
+                                    self.unexpected_illegal_instr_cnt += 1
+                            # TODO: call sample here to do the functional coverage, and move assign_trace_info into it
+                            #self.sample(instruction)
```
Comment (udinator) [inline pygen/pygen_src/test/riscv_instr_cov_test.py:79]: Are these comments necessary, since `self.sample()` is already called above?

Comment (hodjat91) [reply pygen/pygen_src/test/riscv_instr_cov_test.py:79]: Oops, that's already called. I'll fix this one too.

File: pygen/pygen_src/test/riscv_instr_cov_test.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+def parse_args():
+    parse = argparse.ArgumentParser()
+    parse.add_argument("--csv_dir", help="directory consists of the CSV files", type=str, default = ".")
+    args = parse.parse_args()
+    return args
+
+def main():
+    parse = argparse.ArgumentParser()
+    parse.add_argument("--csv_dir", help="directory consists of the CSV files", type=str, default = ".")
+    args = parse.parse_args()
```
Comment (udinator) [inline pygen/pygen_src/test/riscv_instr_cov_test.py:131]: do you mean to call the `parse_args()` defined above instead?

Comment (udinator) [reply pygen/pygen_src/test/riscv_instr_cov_test.py:131]: sorry, I take back this comment actually.

it's actually unnecessary for this coverage test to search for all the `.csv` files that are generated, as the `cov.py` script already gets the pathname to the CSV filenames [here](https://github.com/google/riscv-dv/blob/master/cov.py#L137-L147) and passes them all into the coverage test as cmdline arguments - you should be able to just iterate through `sys.argv` and pull the pathnames from there directly.

File: pygen/pygen_src/isa/riscv_instr_cov.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    MIN_VAL = auto()
+    MAX_VAL = auto()
+    ZERO_VAL = auto()
+
+# dictionary to keep track of reg_name:reg_value throughout the program 
+gpr_state = {}
+
+def get_gpr(reg_name):
+    reg_name = reg_name.upper()
+    # logging.error...
```
Comment (udinator) [inline pygen/pygen_src/isa/riscv_instr_cov.py:36]: maybe a helper function is needed here to convert a string into the corresponding `Enum` member (in this case converting `reg_name` into `riscv_reg_t` Enum type)?
https://docs.python.org/3/library/enum.html#programmatic-access-to-enumeration-members-and-their-attributes

File: pygen/pygen_src/test/riscv_instr_cov_test.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+            logging.error("{} instruction skipped, {} illegal "
+                          "instructions".format(self.skipped_cnt),
+                          self.unexpected_illegal_instr_cnt)
+
+    def post_process_trace(self):
+        pass
+
+    def sample(self):
+        instr_name, binary = "", ""
+        get_val(self.trace["binary"], binary, hexa=1)
```
Comment (udinator) [inline pygen/pygen_src/test/riscv_instr_cov_test.py:115]: maybe we should have a standardized usage of `get_val(...)`? 
Seems like right now the usage depends on the type of variable it returns (e.g. here we pass the `binary` string by reference but below on line 133 we explicitly assigns the return value to `instruction.pc` - maybe we can eliminate the `output_val` argument of `get_val()` and always explicitly assign a return value for clarity?

File: pygen/pygen_src/isa/riscv_instr_cov.py (hunk 4 above)
Comment (hodjat91) [reply pygen/pygen_src/isa/riscv_instr_cov.py:36]: Thanks for the comment! Fixed.

File: pygen/pygen_src/test/riscv_instr_cov_test.py (hunk 5 above)
Comment (hodjat91) [reply pygen/pygen_src/test/riscv_instr_cov_test.py:115]: That makes sense. It's fixed.

File: .idea/.gitignore [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Comment (udinator) [inline .idea/.gitignore:1]: Several `.idea/*` files seem to have been added?

Comment (hodjat91) [reply .idea/.gitignore:1]: Oh this is the .idea directory created by the IDE, fixed it.

File: pygen/pygen_src/test/riscv_instr_cov_test.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+class riscv_instr_cov_test(argv):
+    """ Main class for applying the functional coverage test """
+    def __init__(self):
+        self.trace = {}
+        self.entry_cnt, self.total_entry_cnt, self.skipped_cnt, \
+        self.unexpected_illegal_instr_cnt = 0, 0, 0, 0
+
+    def run_phase(self):
+        logging.info("{} CSV trace files to be "
```
Comment (udinator) [inline pygen/pygen_src/test/riscv_instr_cov_test.py:162]: should we short-circuit and `sys.exit` right away if argv is empty?

Comment (hodjat91) [reply pygen/pygen_src/test/riscv_instr_cov_test.py:162]: Yup, makes sense.

**Pull request merged**